From 31f95cac8645aa69df2f74c97dbf90299a9c6515 Mon Sep 17 00:00:00 2001
From: Lubomir Rintel <lkundrak@v3.sk>
Date: Mon, 2 Mar 2015 18:51:10 +0100
Subject: [PATCH] device: allow vlans on bond and bridges

---
 src/nm-device.c |    7 +++++--
 1 files changed, 5 insertions(+), 2 deletions(-)

diff --git a/src/nm-device.c b/src/nm-device.c
index 9165cf4..e5267f1 100644
--- a/src/nm-device.c
+++ b/src/nm-device.c
@@ -39,6 +39,8 @@
 #include "nm-device.h"
 #include "nm-device-private.h"
 #include "nm-device-ethernet.h"
+#include "nm-device-bond.h"
+#include "nm-device-bridge.h"
 #include "NetworkManagerUtils.h"
 #include "nm-system.h"
 #include "nm-dhcp-manager.h"
@@ -4980,6 +4982,7 @@ nm_device_clear_autoconnect_inhibit (NMDevice *device)
 gboolean
 nm_device_supports_vlans (NMDevice *device)
 {
-	/* At the moment, NM's VLAN code assumes all VLANs are over ethernet. */
-	return NM_IS_DEVICE_ETHERNET (device);
+	return    NM_IS_DEVICE_ETHERNET (device)
+	       || NM_IS_DEVICE_BOND (device)
+	       || NM_IS_DEVICE_BRIDGE (device);
 }
-- 
1.7.1

