diff -up NetworkManager-0.8.1/src/ppp-manager/nm-ppp-manager.c.pppoe-userland NetworkManager-0.8.1/src/ppp-manager/nm-ppp-manager.c
--- NetworkManager-0.8.1/src/ppp-manager/nm-ppp-manager.c.pppoe-userland	2014-04-04 12:52:41.579913527 -0400
+++ NetworkManager-0.8.1/src/ppp-manager/nm-ppp-manager.c	2014-04-04 12:59:52.291909324 -0400
@@ -600,6 +600,27 @@ nm_find_pppd (void)
 	return *pppd_binary;
 }
 
+static inline const char *
+nm_find_pppoe (void)
+{
+	static const char *pppoe_binary_paths[] = {
+		"/usr/local/sbin/pppoe",
+		"/usr/sbin/pppoe",
+		"/sbin/pppoe",
+		NULL
+	};
+
+	const char  **pppoe_binary = pppoe_binary_paths;
+
+	while (*pppoe_binary != NULL) {
+		if (g_file_test (*pppoe_binary, G_FILE_TEST_EXISTS))
+			break;
+		pppoe_binary++;
+	}
+
+	return *pppoe_binary;
+}
+
 static void
 ppp_exit_code (guint pppd_exit_status, GPid pid)
 {
@@ -718,7 +739,7 @@ create_pppd_cmd_line (NMPPPManager *self
                       GError **err)
 {
 	NMPPPManagerPrivate *priv = NM_PPP_MANAGER_GET_PRIVATE (self);
-	const char *ppp_binary;
+	const char *ppp_binary, *pppoe_binary;
 	NMCmdLine *cmd;
 	const char *ppp_debug;
 
@@ -729,6 +750,15 @@ create_pppd_cmd_line (NMPPPManager *self
 		return NULL;
 	}
 
+	if (pppoe) {
+		pppoe_binary = nm_find_pppoe ();
+		if (!pppoe_binary) {
+			g_set_error (err, NM_PPP_MANAGER_ERROR, NM_PPP_MANAGER_ERROR,
+			             "Could not find pppoe binary.");
+			return NULL;
+		}
+	}
+
 	/* Create pppd command line */
 	cmd = nm_cmd_line_new ();
 	nm_cmd_line_add_string (cmd, ppp_binary);
@@ -749,21 +779,30 @@ create_pppd_cmd_line (NMPPPManager *self
 	}
 
 	if (pppoe) {
-		char *dev_str;
+		GString *pppoe_arg;
 		const char *pppoe_service;
+		char *quoted;
 
-		nm_cmd_line_add_string (cmd, "plugin");
-		nm_cmd_line_add_string (cmd, "rp-pppoe.so");
+		g_assert (pppoe_binary != NULL);
+		pppoe_arg = g_string_new (pppoe_binary);
 
-		dev_str = g_strdup_printf ("nic-%s", priv->parent_iface);
-		nm_cmd_line_add_string (cmd, dev_str);
-		g_free (dev_str);
+		g_string_append (pppoe_arg, " -I ");
+		quoted = g_shell_quote (priv->parent_iface);
+		g_string_append (pppoe_arg, quoted);
+		g_free (quoted);
 
 		pppoe_service = nm_setting_pppoe_get_service (pppoe);
 		if (pppoe_service) {
-			nm_cmd_line_add_string (cmd, "rp_pppoe_service");
-			nm_cmd_line_add_string (cmd, pppoe_service);
+			g_string_append (pppoe_arg, " -S ");
+			quoted = g_shell_quote (pppoe_service);
+			g_string_append (pppoe_arg, quoted);
+			g_free (quoted);
 		}
+
+		nm_cmd_line_add_string (cmd, "pty");
+		nm_cmd_line_add_string (cmd, pppoe_arg->str);
+
+		g_string_free (pppoe_arg, TRUE);
 	} else {
 		nm_cmd_line_add_string (cmd, priv->parent_iface);
 		/* Don't send some random address as the local address */
