From 3351e2f70a6dc90137f6ccc3644b986f350b783b Mon Sep 17 00:00:00 2001
From: Lubomir Rintel <lkundrak@v3.sk>
Date: Wed, 4 Mar 2015 18:26:39 +0100
Subject: [PATCH] device: make the activation succeed if no L3 configuration is desired

---
 src/nm-device.c |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/src/nm-device.c b/src/nm-device.c
index e5267f1..b59555e 100644
--- a/src/nm-device.c
+++ b/src/nm-device.c
@@ -2190,6 +2190,10 @@ nm_device_activate_stage3_ip_config_start (gpointer user_data)
 	if (!nm_device_activate_stage3_ip6_start (self))
 		goto out;
 
+	/* If there's no L3 configuration there's no further activation needed. */
+	if (priv->ip4_ready && priv->ip6_ready)
+		nm_device_state_changed (self, NM_DEVICE_STATE_ACTIVATED, NM_DEVICE_STATE_REASON_NONE);
+
 out:
 	nm_log_info (LOGD_DEVICE, "Activation (%s) Stage 3 of 5 (IP Configure Start) complete.", iface);
 	return FALSE;
-- 
1.7.1

