diff -up NetworkManager-0.8.1/system-settings/plugins/ifcfg-rh/reader.c.rh1059698-legacy-slave-compat NetworkManager-0.8.1/system-settings/plugins/ifcfg-rh/reader.c
diff -up NetworkManager-0.8.1/system-settings/plugins/ifcfg-rh/writer.c.rh1059698-legacy-slave-compat NetworkManager-0.8.1/system-settings/plugins/ifcfg-rh/writer.c
--- NetworkManager-0.8.1/system-settings/plugins/ifcfg-rh/writer.c.rh1059698-legacy-slave-compat	2014-11-21 07:56:32.814492534 -0500
+++ NetworkManager-0.8.1/system-settings/plugins/ifcfg-rh/writer.c	2014-11-21 07:58:30.411072896 -0500
@@ -1325,9 +1325,10 @@ write_connection_setting (NMSettingConne
 
 	master = nm_setting_connection_get_master (s_con);
 	if (master) {
-		if (nm_setting_connection_is_slave_type (s_con, NM_SETTING_BOND_SETTING_NAME))
+		if (nm_setting_connection_is_slave_type (s_con, NM_SETTING_BOND_SETTING_NAME)) { 
 			svSetValue (ifcfg, "MASTER", master, FALSE);
-		else if (nm_setting_connection_is_slave_type (s_con, NM_SETTING_BRIDGE_SETTING_NAME))
+			svSetValue (ifcfg, "SLAVE", "yes", FALSE);
+		} else if (nm_setting_connection_is_slave_type (s_con, NM_SETTING_BRIDGE_SETTING_NAME))
 			svSetValue (ifcfg, "BRIDGE", master, FALSE);
 	}
 
