diff -up NetworkManager-0.8.1/network-manager-applet-0.8.1/src/applet.glade.foo NetworkManager-0.8.1/network-manager-applet-0.8.1/src/applet.glade
--- NetworkManager-0.8.1/network-manager-applet-0.8.1/src/applet.glade.foo	2010-08-08 06:23:44.000000000 +0200
+++ NetworkManager-0.8.1/network-manager-applet-0.8.1/src/applet.glade	2011-06-24 12:14:23.911503618 +0200
@@ -597,6 +597,7 @@ Shared Key</property>
                 <property name="visible">True</property>
                 <property name="can_focus">True</property>
                 <property name="visibility">False</property>
+                <property name="activates_default">True</property>
               </widget>
               <packing>
                 <property name="left_attach">1</property>
@@ -628,6 +629,7 @@ Shared Key</property>
               <widget class="GtkEntry" id="leap_username_entry">
                 <property name="visible">True</property>
                 <property name="can_focus">True</property>
+                <property name="activates_default">True</property>
               </widget>
               <packing>
                 <property name="left_attach">1</property>
@@ -866,6 +868,7 @@ Shared Key</property>
               <widget class="GtkEntry" id="eap_tls_identity_entry">
                 <property name="visible">True</property>
                 <property name="can_focus">True</property>
+                <property name="activates_default">True</property>
               </widget>
               <packing>
                 <property name="left_attach">1</property>
@@ -964,6 +967,7 @@ Shared Key</property>
                 <property name="visible">True</property>
                 <property name="can_focus">True</property>
                 <property name="visibility">False</property>
+                <property name="activates_default">True</property>
               </widget>
               <packing>
                 <property name="left_attach">1</property>
@@ -1069,6 +1073,7 @@ Shared Key</property>
                 <property name="visible">True</property>
                 <property name="can_focus">True</property>
                 <property name="visibility">False</property>
+                <property name="activates_default">True</property>
               </widget>
               <packing>
                 <property name="left_attach">1</property>
@@ -1156,6 +1161,7 @@ Shared Key</property>
               <widget class="GtkEntry" id="eap_ttls_anon_identity_entry">
                 <property name="visible">True</property>
                 <property name="can_focus">True</property>
+                <property name="activates_default">True</property>
               </widget>
               <packing>
                 <property name="left_attach">1</property>
@@ -1298,6 +1304,7 @@ Shared Key</property>
                 <property name="visible">True</property>
                 <property name="can_focus">True</property>
                 <property name="visibility">False</property>
+                <property name="activates_default">True</property>
               </widget>
               <packing>
                 <property name="left_attach">1</property>
@@ -1311,6 +1318,7 @@ Shared Key</property>
               <widget class="GtkEntry" id="eap_simple_username_entry">
                 <property name="visible">True</property>
                 <property name="can_focus">True</property>
+                <property name="activates_default">True</property>
               </widget>
               <packing>
                 <property name="left_attach">1</property>
@@ -1673,6 +1681,7 @@ Shared Key</property>
               <widget class="GtkEntry" id="eap_peap_anon_identity_entry">
                 <property name="visible">True</property>
                 <property name="can_focus">True</property>
+                <property name="activates_default">True</property>
               </widget>
               <packing>
                 <property name="left_attach">1</property>
