From ee08b7f875866f3fa35a2c621eeac3777ee27685 Mon Sep 17 00:00:00 2001
From: Jiří Klimeš <jklimes@redhat.com>
Date: Tue, 15 Mar 2011 12:44:29 +0000
Subject: gconf: fix accessing freed memory

---
diff --git a/network-manager-applet-0.8.1/src/gconf-helpers/gconf-helpers.c b/network-manager-applet-0.8.1/src/gconf-helpers/gconf-helpers.c
index 19ca925..17b8c24 100644
--- a/network-manager-applet-0.8.1/src/gconf-helpers/gconf-helpers.c
+++ b/network-manager-applet-0.8.1/src/gconf-helpers/gconf-helpers.c
@@ -762,10 +762,10 @@ out:
 
 gboolean
 nm_gconf_get_ip6addr_array_helper (GConfClient *client,
-								   const char *path,
-								   const char *key,
-								   const char *setting,
-								   GPtrArray **value)
+                                   const char *path,
+                                   const char *key,
+                                   const char *setting,
+                                   GPtrArray **value)
 {
 	char *gc_key;
 	GConfValue *gc_value = NULL;
@@ -818,16 +818,17 @@ nm_gconf_get_ip6addr_array_helper (GConfClient *client,
 			g_free (addr);
 			continue;
 		}
-		g_free (addr);
 
 		memset (&rawgw, 0, sizeof (rawgw));
 		if (gw) {
 			if (inet_pton (AF_INET6, gw, &rawgw) <= 0) {
 				g_warning ("%s: %s contained bad gateway address: %s",
 						   __func__, gc_key, gw);
+				g_free (addr);
 				continue;
 			}
 		}
+		g_free (addr);
 
 		valarr = g_value_array_new (3);
 
--
cgit v0.9
