diff -up NetworkManager-0.8.1/libnm-glib/nm-device.c.foo NetworkManager-0.8.1/libnm-glib/nm-device.c
--- NetworkManager-0.8.1/libnm-glib/nm-device.c.foo	2011-08-15 16:57:40.326684085 -0500
+++ NetworkManager-0.8.1/libnm-glib/nm-device.c	2011-08-15 16:57:50.412557995 -0500
@@ -689,7 +689,6 @@ nm_device_new (DBusGConnection *connecti
 						    G_TYPE_STRING, "DeviceType",
 						    G_TYPE_INVALID,
 						    G_TYPE_VALUE, &value, G_TYPE_INVALID)) {
-		g_warning ("Error in get_property: %s\n", err->message);
 		g_error_free (err);
 		goto out;
 	}
