--- a/network-manager-applet-0.8.1/src/applet-device-wifi.c
+++ b/network-manager-applet-0.8.1/src/applet-device-wifi.c
@@ -1342,8 +1342,18 @@ wireless_get_icon (NMDevice *device,
 static void
 activate_device_cb (gpointer user_data, const char *path, GError *error)
 {
-	if (error)
-		nm_warning ("Device Activation failed: %s", error->message);
+	if (error) {
+		char *text = g_strdup_printf ("Activation failed: %s", error->message ? error->message : "unknown");
+
+		nm_warning ("Device %s", text);
+		applet_do_notify (NM_APPLET (user_data),
+		                  NOTIFY_URGENCY_LOW,
+		                  "Connection activation failure",
+		                  text,
+		                  "gnome-lockscreen",
+		                  NULL, NULL, NULL, NULL);
+		g_free (text);
+	}
 	applet_schedule_update_icon (NM_APPLET (user_data));
 }
 
