From 4501e33b852be184a30cbdc2e49803d341db9d0c Mon Sep 17 00:00:00 2001
From: Jiří Klimeš <jklimes@redhat.com>
Date: Mon, 01 Nov 2010 15:02:32 +0000
Subject: dispatcher: don't warn about invalid connection for 'hostname' action (rh #627649)

---
diff --git a/callouts/nm-dispatcher-action.c b/callouts/nm-dispatcher-action.c
index 828d8be..8040853 100644
--- a/callouts/nm-dispatcher-action.c
+++ b/callouts/nm-dispatcher-action.c
@@ -489,6 +489,10 @@ nm_dispatcher_action (Handler *h,
 	if (!d->persist)
 		d->quit_timeout = g_timeout_add_seconds (10, quit_timeout_cb, NULL);
 
+	/* Hostname changes don't require a device nor contain a connection */
+	if (!strcmp (action, "hostname"))
+		goto dispatch;
+
 	connection = nm_connection_new_from_hash (connection_hash, error);
 	if (connection) {
 		NMSettingConnection *s_con;
@@ -506,10 +510,6 @@ nm_dispatcher_action (Handler *h,
 		*error = NULL;
 	}
 
-	/* Hostname changes don't require a device */
-	if (!strcmp (action, "hostname"))
-		goto dispatch;
-
 	/* interface name */
 	value = g_hash_table_lookup (device_props, NMD_DEVICE_PROPS_INTERFACE);
 	if (!value || !G_VALUE_HOLDS_STRING (value)) {
--
cgit v0.9.0.2-2-gbebe
