diff -up NetworkManager-0.8.1/src/dhcp-manager/nm-dhcp-client.c.foo NetworkManager-0.8.1/src/dhcp-manager/nm-dhcp-client.c
--- NetworkManager-0.8.1/src/dhcp-manager/nm-dhcp-client.c.foo	2012-05-16 03:25:16.633082673 -0500
+++ NetworkManager-0.8.1/src/dhcp-manager/nm-dhcp-client.c	2012-05-16 03:26:06.300891993 -0500
@@ -572,8 +572,15 @@ nm_dhcp_client_new_options (NMDHCPClient
 	g_hash_table_remove_all (priv->options);
 	g_hash_table_foreach (options, copy_option, priv->options);
 
-	if (old_state == new_state)
-		return;
+	if (old_state == new_state) {
+		if (   new_state != DHC_BOUND4
+		    && new_state != DHC_RENEW4
+		    && new_state != DHC_REBIND4
+		    && new_state != DHC_BOUND6
+		    && new_state != DHC_RENEW6
+		    && new_state != DHC_REBIND6)
+			return;
+	}
 
 	/* Handle changed device state */
 	if (state_is_bound (new_state)) {

