--- NetworkManager-0.8.1/src/dhcp-manager/nm-dhcp-dhclient.c	2013-03-08 20:49:34.037199894 +0100
+++ NetworkManager-0.8.1/src/dhcp-manager/nm-dhcp-dhclient.c.foo	2013-03-19 15:25:10.150910732 +0100
@@ -509,6 +509,7 @@
 	/* Everything we want to request from the DHCP server */
 	add_also_request (alsoreq, "rfc3442-classless-static-routes");
 	add_also_request (alsoreq, "ms-classless-static-routes");
+	add_also_request (alsoreq, "static-routes");
 	add_also_request (alsoreq, "ntp-servers");
 
 	/* And add it to the dhclient configuration */
