--- NetworkManager-0.8.1/network-manager-applet-0.8.1/src/connection-editor/nm-connection-editor.c	2013-07-04 12:26:07.600623542 +0200
+++ NetworkManager-0.8.1/network-manager-applet-0.8.1/src/connection-editor/nm-connection-editor.c.foo	2013-07-04 12:27:04.615910752 +0200
@@ -207,7 +207,7 @@ update_sensitivity (NMConnectionEditor *
 	gtk_widget_set_sensitive (widget, sensitive);
 
 	widget = GTK_WIDGET (gtk_builder_get_object (editor->builder, "connection_autoconnect"));
-	gtk_widget_set_sensitive (widget, sensitive && !nm_setting_connection_get_master (s_con));
+	gtk_widget_set_sensitive (widget, sensitive);
 
 	widget = GTK_WIDGET (gtk_builder_get_object (editor->builder, "connection_name"));
 	gtk_widget_set_sensitive (widget, sensitive);
