--- NetworkManager-0.8.1/network-manager-applet-0.8.1/src/connection-editor/page-bond.c	2013-07-09 13:20:53.645654948 +0200
+++ NetworkManager-0.8.1/network-manager-applet-0.8.1/src/connection-editor/page-bond.c.foo	2013-07-09 13:21:55.800877900 +0200
@@ -285,6 +285,22 @@ connection_added (NMSettingsInterface *s
 }
 
 static void
+bonding_mode_changed (GtkComboBox *combo, gpointer user_data)
+{
+	CEPageBond *self = user_data;
+	CEPageBondPrivate *priv = CE_PAGE_BOND_GET_PRIVATE (self);
+
+	/* balance-tlb and balance-alb work only with MII monitoring */
+	if (   gtk_combo_box_get_active (combo) == MODE_BALANCE_TLB
+	    || gtk_combo_box_get_active (combo) == MODE_BALANCE_ALB) {
+		gtk_combo_box_set_active (priv->monitoring, MONITORING_MII);
+		gtk_widget_set_sensitive (GTK_WIDGET (priv->monitoring), FALSE);
+	} else {
+		gtk_widget_set_sensitive (GTK_WIDGET (priv->monitoring), TRUE);
+	}
+}
+
+static void
 monitoring_mode_changed (GtkComboBox *combo, gpointer user_data)
 {
 	CEPageBond *self = user_data;
@@ -451,6 +467,10 @@ populate_ui (CEPageBond *self)
 			mode_idx = MODE_BALANCE_ALB;
 	}
 	gtk_combo_box_set_active (priv->mode, mode_idx);
+	g_signal_connect (priv->mode, "changed",
+	                  G_CALLBACK (bonding_mode_changed),
+	                  self);
+	bonding_mode_changed (priv->mode, self);
 
 	/* Monitoring mode/frequency */
 	frequency = nm_setting_bond_get_option_by_name (setting, NM_SETTING_BOND_OPTION_ARP_INTERVAL);
@@ -852,6 +872,9 @@ ui_to_setting (CEPageBond *self)
 	downdelay = gtk_entry_get_text (GTK_ENTRY (priv->downdelay));
 	targets = uglify_targets (gtk_entry_get_text (priv->arp_targets));
 
+	/* Set bond mode */
+	nm_setting_bond_add_option (priv->setting, NM_SETTING_BOND_OPTION_MODE, mode);
+
 	switch (gtk_combo_box_get_active (priv->monitoring)) {
 	case MONITORING_MII:
 		nm_setting_bond_add_option (priv->setting, NM_SETTING_BOND_OPTION_MIIMON, frequency);
