diff --git a/src/system-settings/nm-sysconfig-settings.c b/src/system-settings/nm-sysconfig-settings.c
index 36dd582..edff6bb 100644
--- a/src/system-settings/nm-sysconfig-settings.c
+++ b/src/system-settings/nm-sysconfig-settings.c
@@ -400,6 +400,27 @@ find_plugin (GSList *list, const char *pname)
 	return obj;
 }
 
+static char **
+add_keyfile_plugin (char **plugins)
+{
+	char **new_plugins;
+	int i;
+
+	for (i = 0; plugins[i]; i++) {
+		if (!strcmp (plugins[i], "keyfile"))
+			return plugins;
+	}
+
+	new_plugins = g_new (char *, i + 2);
+	for (i = 0; plugins[i]; i++)
+		new_plugins[i] = plugins[i];
+	new_plugins[i] = g_strdup ("keyfile");
+	new_plugins[i + 1] = NULL;
+
+	g_free (plugins);
+	return new_plugins;
+}
+
 static gboolean
 load_plugins (NMSysconfigSettings *self, const char *plugins, GError **error)
 {
@@ -411,6 +432,7 @@ load_plugins (NMSysconfigSettings *self, const char *plugins, GError **error)
 	plist = g_strsplit (plugins, ",", 0);
 	if (!plist)
 		return FALSE;
+	plist = add_keyfile_plugin (plist);
 
 	for (iter = plist; *iter; iter++) {
 		GModule *plugin;
