--- NetworkManager-0.8.1/src/nm-netlink-monitor.c      2013-07-09 13:50:31.503428667 +0200
+++ NetworkManager-0.8.1/src/nm-netlink-monitor.c.foo  2013-07-09 13:50:26.954485536 +0200
@@ -39,7 +39,6 @@
 #include <linux/rtnetlink.h>
 #include <linux/if.h>
 #include <linux/if_bonding.h>
-#include <linux/if_bridge.h>
 #include <linux/sockios.h>
 #include <linux/unistd.h>
 #include <unistd.h>
--- NetworkManager-0.8.1/src/nm-system.c	2013-07-09 13:20:53.858652285 +0200
+++ NetworkManager-0.8.1/src/nm-system.c.foo	2013-07-09 13:27:11.872926456 +0200
@@ -44,7 +44,6 @@
 #include <linux/if.h>
 #include <linux/sockios.h>
 #include <linux/if_bonding.h>
-#include <linux/if_bridge.h>
 
 #include "nm-system.h"
 #include "nm-device.h"
